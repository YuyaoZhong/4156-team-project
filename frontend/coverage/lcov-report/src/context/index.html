
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src/context</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> src/context</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">38.14% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>74/194</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">9.38% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>6/64</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">30.19% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>16/53</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">37.5% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>66/176</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
    </div>
    <div class='status-line low'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="data-context.js"><a href="data-context.js.html">data-context.js</a></td>
	<td data-value="37.23" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 37%"></div><div class="cover-empty" style="width: 63%"></div></div>
	</td>
	<td data-value="37.23" class="pct low">37.23%</td>
	<td data-value="188" class="abs low">70/188</td>
	<td data-value="9.38" class="pct low">9.38%</td>
	<td data-value="64" class="abs low">6/64</td>
	<td data-value="29.41" class="pct low">29.41%</td>
	<td data-value="51" class="abs low">15/51</td>
	<td data-value="36.84" class="pct low">36.84%</td>
	<td data-value="171" class="abs low">63/171</td>
	</tr>

<tr>
	<td class="file medium" data-value="google-login-context.js"><a href="google-login-context.js.html">google-login-context.js</a></td>
	<td data-value="66.67" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 66%"></div><div class="cover-empty" style="width: 34%"></div></div>
	</td>
	<td data-value="66.67" class="pct medium">66.67%</td>
	<td data-value="6" class="abs medium">4/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="2" class="abs medium">1/2</td>
	<td data-value="60" class="pct medium">60%</td>
	<td data-value="5" class="abs medium">3/5</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank">istanbul</a>
                at Sat Dec 05 2020 18:31:18 GMT-0800 (GMT-08:00)
            </div>
        </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    