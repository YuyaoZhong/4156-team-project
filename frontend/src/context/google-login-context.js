import React from 'react';
import { useGoogleLogin } from 'react-use-googlelogin';
import { GOOGLE_CLIENT_ID } from '../constants/constants';

export const GoogleAuthContext = React.createContext();

// mark: change to process.env will be better
export const GoogleAuthProvider = ({ children }) => {
    const googleAuth = useGoogleLogin({
        clientId: GOOGLE_CLIENT_ID
    });

    return (<GoogleAuthContext.Provider value={googleAuth}>
        {children}
    </GoogleAuthContext.Provider>);
}

export const useGoogleAuth = () => React.useContext(GoogleAuthContext);