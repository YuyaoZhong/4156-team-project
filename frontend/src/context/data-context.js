import React from 'react'
import { SERVER_URL } from '../constants/constants';
import { useGoogleAuth } from './google-login-context';
import { upsertData, deleteData } from '../utilities/apiMethods';

export const DataContext = React.createContext()


const timerSort = (a, b)=> (new Date(a.startTime) - new Date(b.startTime))
const getEndTime = (timer) => (new Date(new Date(timer.startTime).getTime() + (timer.duration + timer.breakTime) * timer.round * 60000));

const getIncomingTimer = (timerlist) => {return timerlist.filter(item=>getEndTime(item).getTime() - new Date().getTime() > 0).sort(timerSort)};


export const DataContextProvider = props => {
    const [timerList, setTimerList] = React.useState([]);
    const [tasks, setTasks] = React.useState([]);
    const [tasklists, setTasklists] = React.useState([]);
    const [timerRun, setTimerRun] = React.useState({});
    // const [taskAndTimer, setTaskToTimer] = React.useState({});
    const [loading, setLoading] = React.useState(false);
    const [curTime, setCurTime] =  React.useState(new Date().getTime());
    const [incomingTimers, setIncomingTimers] = React.useState([]);
    

    const {isSignedIn, googleUser} = useGoogleAuth();
    const userId = isSignedIn ? googleUser.googleId : ""

    React.useEffect(()=>{
    setLoading(true);
       async function fetchData () {
            // const getAllTimerRoute = `${SERVER_URL}/timers/?userId=${userId}`;
            const getAllTimerRoute = `${SERVER_URL}/timerToUser/?userId=${userId}`;
            const getAllTaskRoute = `${SERVER_URL}/tasks?userId=${userId}`;
            // const getAllTasklistRoute = `${SERVER_URL}/tasklists?userId=${userId}`;
            const getAllTasklistRoute =  `${SERVER_URL}/tasklists?userId=${userId}`;
            // const getAllTaskandTimerRoute = `${SERVER_URL}/task_timers?userId=${userId}`;
            const urls = [getAllTimerRoute, getAllTaskRoute, getAllTasklistRoute];
            const promises = urls.map(url=>fetch(url).then(r=>{
                if(r.status === 200){
                    return r.json()
                }
                return {"data": []}
            }));
             //.then(r=>JSON.parse(r)));
            try {
                await Promise.all(promises).then(res=>{
                    console.log(res);
                    const timers = res[0]["data"].sort(timerSort);
                    setIncomingTimers(getIncomingTimer(timers));
                    setTimerList(timers);
                    setTasks(res[1]["data"]);
                    setTasklists(res[2]["data"]);
                })
            } catch(e){
                console.log(e);
            }
            setLoading(false);
       }

       fetchData();
    },[userId])

    const checkTImerRunning = () => {
        // may need to change fetching first

        const curTime = new Date();
        const incomingTimerList = getIncomingTimer(timerList);

        if (Object.keys(timerRun).length !== 0) {
            const lastMins = timerRun.round * (timerRun.duration + timerRun.breakTime);
            const endTime = new Date(new Date(timerRun.startTime).getTime() + lastMins * 60000);
            if (curTime.getTime() - endTime.getTime() >= 0){
                setTimerRun({});
            }
        }

        if (timerList.length > 0){
            
            // const incomingTimerList = timerList.filter(item=>getEndTime(item).getTime() - new Date().getTime() > 0).sort(timerSort);
            const firstTimer = Object.assign({}, incomingTimerList[0]);
            console.log(firstTimer);
            const nextStartTime = new Date(firstTimer.startTime);
            const differTime = nextStartTime.getTime() - curTime.getTime(); // millseconds
      
            // if (differTime < 0.5 * 60000){ // less than one minutes
            //     setTimerRun(firstTimer);
            // }
            // test: change to the first timer
            if(firstTimer.id !== timerRun.id){
                setTimerRun(firstTimer);
            }
        }

        setIncomingTimers(setIncomingTimers);
     
    }

    // set interval for updating running timer
    React.useEffect(()=>{
        const intervalTime = 10000;
        const interval = setInterval(()=>{
          setCurTime(prev=>prev + intervalTime)
        }, intervalTime)
        setIncomingTimers(getIncomingTimer(timerList))
        return ()=>clearInterval(interval);
    }, [timerList]);


    React.useEffect(()=>{
    //    console.log('TO RUN CHECK')
      checkTImerRunning();
    }, [curTime, timerList])

    const updateTimerListState = (timerId, newTimer, edit)=>{
        const updateCallBack = (state) =>{
            const newState = [...state];
            let idx = -1;
            if(edit){
                idx = newState.findIndex(item=>String(item.id) === String(timerId));
            }
            if (idx === -1){
                newState.push(newTimer);
            } else {
                newState[idx] = newTimer;
            }
           
            newState.sort((a, b)=> (new Date(a.startTime) - new Date(b.startTime)));
            // console.log('in add timer', newState)
            return newState;
        }
        setTimerList(updateCallBack);
    }

    const handleCreateTimer = async (timerData, edit) => {
        // console.log(timerData);
        timerData.userId = userId;
        let timerId = timerData.id;
        const route = edit? `${SERVER_URL}/timers/${timerData.id}` : `${SERVER_URL}/timers/`;
        const method = edit ? 'PUT' : 'POST';
        await upsertData(route, timerData, method).then(res=>{
            console.log('timer update', res);
            if(res.code === 201 && res.data){
                timerId = res.data.id;
                updateTimerListState(timerData.id, res.data, edit);
            }
        })

        return timerId; // for redirect
    }


    const handleDeleteTimer = async (timerIds) => {
    //    const deletePromises = timerIds.map(timerId=>(`${SERVER_URL}/timers/${timerId}`));
        const deletePromises = timerIds.map(timerIdObj=>{
            let route = timerIdObj.isCreator?
            `${SERVER_URL}/timers/${timerIdObj.timerId}`
            :`${SERVER_URL}/timerToUser/${timerIdObj.timerId}/${timerIdObj.timerToUserId}`;
            // console.log(route)
            return deleteData(route);
        });


        await Promise.all(deletePromises).then(results=>{

            const deleteIds = results.reduce((res, item)=>{
                if(item.code === 200 && item.data){
                    if(item.data.isTimerToUser){
                        res.push(item.data.timerId);
                    }else{
                        res.push(item.data.id);
                    }
                }
                return res;
            }, []);

            setTimerList(state=>{
                const newState = state.filter(item=>{
                    const idx = deleteIds.findIndex(deleteId=>(String(deleteId) === String(item.id)));
                    return idx === -1; // not delete
                })
                return newState;
            })
        })

    }

    const handleUpsertTask = async (taskData, edit) => {
        taskData.userId = userId;
        if(taskData.taskListId === 0){
            taskData.taskListId = null;
        }
        const route = edit? `${SERVER_URL}/tasks/${taskData.id}`:`${SERVER_URL}/tasks`;
        const method = edit ? 'PUT' : 'POST';
        await upsertData(route, taskData, method).then(res=>{
            console.log('in upsert task', res)
         if(res.code === 201 && res.data){
             console.log(res)
             setTasks(state=>{
                 const newState = [...state];
                 let idx = -1;
                 if(edit){
                   idx = newState.findIndex(item=>(String(item.id) === String(taskData.id)));
                 }
                 if (idx === -1){
                    newState.push(res.data);
                 } else {
                     newState[idx] = res.data;
                 }
                 return newState;
             });
         }
        })
    }

    const getRelatedTasksOfTimers  = async (timerId) => {

        const tasksOftimersRoute = `${SERVER_URL}/task_timers?userId=${userId}&timerId=${timerId}`;
        // console.log(tasksOftimersRoute);
        let relatedTasks = []
        await fetch(tasksOftimersRoute).then(r=>r.json()).then(
            res=>{
                if(res.code === 200){
                    relatedTasks = res.data;
                }
                console.log(res);
            })
        return relatedTasks;
    }
    
    const handleDeleteTask = async (taskData) => {
        taskData.userId = userId;
        const route = `${SERVER_URL}/tasks/${taskData.id}`;
        const method = 'DELETE';
        await upsertData(route, taskData, method).then(res=>{
            console.log('in upsert task', res)
         if(res.code === 201){
             setTasks(state=>{
                 // todo: sort default by incomplete / compete
                 let idx = -1;
                 idx = state.findIndex(item=>(String(item.id) === String(taskData.id)));
                 state.splice(idx,1)
                 const newState = state.splice(0)
                 return newState;
             });
         }
        })
    }

    const handleDeleteTaskList = async (taskListData) => {
        taskListData.userId = userId;
        const route = `${SERVER_URL}/tasklists/${taskListData.id}`;
        const method = 'DELETE';
        await upsertData(route, taskListData, method).then(res=>{
            console.log('in upsert task', res)
         if(res.code === 201){
             setTasklists(state=>{
                 let idx = -1;
                 idx = state.findIndex(item=>(String(item.id) === String(taskListData.id)));
                 state.splice(idx,1)
                 const newState = state.splice(0)
                 return newState;
             });
         }
        })
    }

    const handleUpsertTaskList = async (taskListData, edit) => {
        taskListData.userId = userId;
        const route = edit? `${SERVER_URL}/tasklists/${taskListData.id}`:`${SERVER_URL}/tasklists`;
        const method = edit ? 'PUT' : 'POST';
        await upsertData(route, taskListData, method).then(res=>{
            // console.log('in upsert taskList', res)
        if(res.code === 201 && res.data){
            setTasklists(state=>{
                // todo: sort default by incomplete / compete
                const newState = state.splice(0);
                let idx = -1;
                if(edit){
                idx = newState.findIndex(item=>(String(item.id) === String(taskListData.id)));
                }
                if (idx === -1){
                    newState.push(res.data);
                } else {
                    newState[idx] = res.data;
                }
                return newState;
                });
            }
        })
    }
   

    const getTimerById = (timerId) => {
        const targetTimer = timerList.find(timer=>String(timer.id) === String(timerId));
        return targetTimer;
    }


    return (<DataContext.Provider value = {{
            tasks,
            tasklists,
            timerList,
            incomingTimers,
            loading,
            timerRun,
            getTimerById,
            handleCreateTimer,
            handleUpsertTask,
            getRelatedTasksOfTimers,
            handleDeleteTimer,
            handleUpsertTaskList,
            handleDeleteTask,
            handleDeleteTaskList,
            updateTimerListState,
            userId,
    }}>
        {props.children}
    </DataContext.Provider>)
}

export const useDataContext = () => React.useContext(DataContext);
