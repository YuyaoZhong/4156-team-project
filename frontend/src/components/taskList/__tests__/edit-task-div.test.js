import React from 'react'
import 'jest-enzyme';
import Enzyme, { shallow, mount} from 'enzyme'
import { configure } from "enzyme";
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import renderer from 'react-test-renderer';
import EditTaskDiv from '../edit-task-div';
import { DataContext} from '../../../context/data-context';


configure({ adapter: new Adapter() });

describe("test <EditTaskDiv/>", () => {
    let mockHandleClose;
    let wrapper;

    beforeEach(() => {
        mockHandleClose = jest.fn();
        wrapper = mount(<DataContext.Provider value = {{
            handleUpsertTask: jest.fn()
         }}
        >
            <EditTaskDiv closeAddTaskMode={mockHandleClose}/>
        </DataContext.Provider>)
        // wrapper = shallow(<EditTaskDiv closeAddTaskMode={mockHandleClose}/>);
      });

    it("test an valid name", async () => {
       const testName = "Test Task";
       wrapper.find('input[name="task name"]').simulate('change', {
           target:{ 
               value: testName
           }
       });

       expect(wrapper.find('input[name="task name"]').prop('value')).toEqual(testName);
    });

    it("test a boundary name", async () => {
        let testBoundaryName = "";
        const boundarylength = 140;
        for(var i = 0; i < boundarylength; i++){
            testBoundaryName += "a";
        }

        expect(testBoundaryName.length).toEqual(boundarylength);

        wrapper.find('input[name="task name"]').simulate('change', {
            target:{ 
                value: testBoundaryName
            }
        });

        expect(wrapper.find('input[name="task name"]').prop('value')).toEqual(testBoundaryName);
    });


    it("test a too long name", async () => {
        let testTooLongName = "";
        const boundarylength = 140;
        const tooLong = 200;
        for(var i = 0; i < tooLong; i++){
            testTooLongName += "a";
        }

        expect(testTooLongName.length).toEqual(tooLong);

        wrapper.find('input[name="task name"]').simulate('change', {
            target:{ 
                value: testTooLongName
            }
        });

        expect(wrapper.find('input[name="task name"]').prop('value').length).toEqual(boundarylength);
    })


    it("test submit", ()=>{
    
        wrapper.find("button").first().simulate('click');
        expect(mockHandleClose).toHaveBeenCalled();

    })
    
    it("snapshot tests", () => {
        const editTaskDivRender = renderer.create((<DataContext.Provider value = {{
            handleUpsertTask: jest.fn()
         }}
        >
            <EditTaskDiv closeAddTaskMode={mockHandleClose}/>
        </DataContext.Provider>)).toJSON();

        expect(editTaskDivRender).toMatchSnapshot();
    })


  });