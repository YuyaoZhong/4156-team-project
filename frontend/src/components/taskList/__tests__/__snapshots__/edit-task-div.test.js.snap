// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <EditTaskDiv/> snapshot tests 1`] = `
<form
  className="ui large form"
  onSubmit={[Function]}
>
  <div
    className="field"
  >
    <label>
      name
    </label>
    <input
      aria-describedby={null}
      maxLength={140}
      name="task name"
      onChange={[Function]}
      type="text"
      value="New Task"
    />
  </div>
  <button
    className="ui primary button"
    onClick={[Function]}
    type="button"
  >
    Save 
  </button>
  <button
    className="ui secondary button"
    onClick={[Function]}
    type="button"
  >
    Cancel
  </button>
</form>
`;
